### PR TITLE
SmartPort - Some fixes

### PR DESCRIPTION
**Pull request by kilrah**
- Suppress messages for disabled features and unavailable sensors (save bandwidth and clutter/storage on OpenTX 2.1)
- Fix stray responses to outdated requests from the receiver (sometimes especially shortly after boot the board would be too slow to service incoming serial data, and subsequent parsing of the buffer would trigger a request when the receiver had already sent IDs other than the one the board is supposed to answer to, leading to the board stepping over someone else's slot and being detected as a new sensor).
- A smart port request is only valid for a few ms, so the current way of looping through the ID list was not valid. One solution would be to just discard the slot and wait for the next one, the other is to loop straight away while we're there. I implemented the latter as the former would cause a loss of bandwidth and the code seems light enough to me not to cause an unacceptable delay, but feel free to correct me if I'm mistaken.
- I removed the service delay that IMO is useless, it was checking that we don't answer twice in <5ms to a request that can't come more frequently than every 12ms in theory anyway (24ms in practice). Again feel free to object and discard the commit if it was only meant as a safety precaution.

Everything now seems to play nice and within spec, including when connected to the same bus with other sensors.
Good job on the implementation BTW, I overhauled my mini H that was still on Baseflight a few weeks ago and wired the smart port to SOFTSERIAL1 without hardware inverter and together with my 40A smart port FrSky current sensor in the way I *thought* it should be good to make it work hoping to get to the implementation at some point, but had no time. When I came back and flashed 1.8.1 it just worked.

screenshot-2015-04-26-19-07-55

screenshot-2015-04-26-19-07-58

### CODE REVIEW

**Comment by hydra** [comment]
@kilrah thanks for the updates and fixes, much appreciated.  Got quite a few things to merge in at the moment, will look at this soon.

File: src/main/telemetry/smartport.c [hunk 1]
```diff
@@ -161,8 +160,8 @@ static void smartPortDataReceive(uint16_t c)
     static uint8_t lastChar;
     if (lastChar == FSSP_START_STOP) {
         smartPortState = SPSTATE_WORKING;
-        smartPortLastRequestTime = now;
-        if (c == FSSP_SENSOR_ID1) {
+        if (c == FSSP_SENSOR_ID1 && (serialTotalBytesWaiting(smartPortSerialPort) == 0)) {
```
Comment (ledvinap) [inline src/main/telemetry/smartport.c]: `serialTotalBytesWaiting` wont help here - when using callback, input buffer is skipped completely ...

**Comment by Netsplits** [comment]
I've made some tests with this PR and seems to work as expected. I've logic analysed the serial data BEFORE using this patch and I've seen outdated responses to wrong IDs (0xA1 instead of 0x1B) as kilrah says.

Please forgive me if I'm saying something stupid, but how many iterations in the while loop will the cpu do before timeouts with only 1ms timeout? I mean, with that low value will the while be executed more than 1 time?

Kinda offtopic: @kilrah You say connect the smartport to softserial 1. Is the smartport signal voltage 3.3v? Or are the SOFT SERIAL 1 pins on the Naze32 5V tolerant?

**Comment by kilrah** [comment]
> Please forgive me if I'm saying something stupid, but how many iterations in the while loop will the cpu do before timeouts with only 1ms timeout? I mean, with that low value will the while be executed more than 1 time?

I'd estimate a loop in there not to take more than a dozen of microseconds or so. The timeout is really just a safety, as there is always at least a value being sent there is no reason for it to trigger... but as it would have the potential to hang everything I felt safer to leave it just in case a future change screwed it up.

> Is the smartport signal voltage 3.3v

Yes.

**Comment by Netsplits** [comment]
But smartport speed is 57600 bauds, is softserial already fast enough? Docs say that 19200 bauds is the max speed of softserials...

**Comment by kilrah** [comment]
Well as it works perfectly I'd say yes :)

**Comment by Netsplits** [comment]
Doh! I made a hardware inverter to use telemetry on UART 1 (shared with MSP)... Oh well, good to know, this way I'll be able to connect the OSD on softserial if I ever buy one. Thanks for the info!

**Comment by hydra** [comment]
@Netsplits thanks for the feedback

**Pull request merged**